src/opf/ptdf.jl

```diff
@@ -0,0 +1,122 @@
+using SparseArrays
+using PowerModels
+using LinearAlgebra
+
+abstract type AbstractPTDF end
+
+
+struct LazyPTDF{TF} <: AbstractPTDF
+    N::Int  # number of buses
+    E::Int  # number of branches
+    islack::Int  # Index of slack bus
+
+    A::SparseMatrixCSC{Float64,Int}  # incidence matrix
+    b::Vector{Float64}  # branch susceptances
+    BA::SparseMatrixCSC{Float64,Int}  # B*A
+    AtBA::SparseMatrixCSC{Float64,Int}  # AᵀBA
+
+    F::TF   # Factorization of -(AᵀBA). Must be able to solve linear systems with F \ p
+            # We use a factorization of -(AᵀBA) to support cholesky factorization when possible
+
+    # TODO: cache
+end
+
+function LazyPTDF(data::OPFData; solver::Symbol=:ldlt)
+    N, E, A, b, ref_idx = data.N, data.E, data.A, data.b, data.ref_bus
+
+    B = Diagonal(b)
+    BA = B * A
+    S = AtBA = A' * BA
+
+    S[ref_idx, :] .= 0.0
+    S[:, ref_idx] .= 0.0
+    S[ref_idx, ref_idx] = -1.0;  # to enable cholesky
+    S = -S
+
+    if solver == :ldlt
+        F = ldlt(S)
+    elseif solver == :cholesky
+        # If Cholesky is not possible, default to LDLᵀ
+        if maximum(b) < 0.0
+            F = cholesky(S)
+        else
+            @warn "Some branches have positive susceptance, cannot use Cholesky; defaulting to LDLᵀ"
+            F = ldlt(S)
+        end
+    else
+        error("Invalid linear solver: only ldlt and cholesky are supported")
+    end
+
+    return LazyPTDF(N, E, ref_idx, A, b, BA, AtBA, F)
+end
```

Comment:
How about we drop Cholesky and keep only `LDLt`? 
* Supporting both Cholesky and LDLt introduce an additional argument and type parameter, plus the possibility of a runtime error
* Memory-wise, LDLt and Cholesky are basically identical
* Cholesky is valid only if all branch resistances have the same sign, whereas LDLt works all the time
* To support Cholesky in a unified way, the initial implementation negated the matrix `S`... which increases the risk of a downstream bug

If we do, then here is a summary of changes:
* Remove type parametrization of `F` in the struct --> use the type of an LDLt factorization directly
* Remove `solver` argument
* Replace `S[ref_idx, ref_idx] = -1.0` with `+1.0` and delete inline comment
* Remove `S = -S`
* Remove the `if...else` statement checking for `solver` value
* Update sign conventions in downstream functions